src/utils/resolve_url.ts

```diff
@@ -14,80 +14,26 @@
  * limitations under the License.
  */
 
+import startsWith from "./starts_with";
+
 // Scheme part of an url (e.g. "http://").
 const schemeRe = /^(?:[a-z]+:)?\/\//i;
 
-// Captures "/../" or "/./".
-const selfDirRe = /\/\.{1,2}\//;
-
-/**
- * Resolve self directory and previous directory references to obtain a
- * "normalized" url.
- * @example "https://foo.bar/baz/booz/../biz" => "https://foo.bar/baz/biz"
- * @param {string} url
- * @returns {string}
- */
-function _normalizeUrl(url: string): string {
-  // fast path if no ./ or ../ are present in the url
-  if (!selfDirRe.test(url)) {
-    return url;
-  }
-
-  const newUrl: string[] = [];
-  const oldUrl = url.split("/");
-  for (let i = 0, l = oldUrl.length; i < l; i++) {
-    if (oldUrl[i] === "..") {
-      newUrl.pop();
-    } else if (oldUrl[i] === ".") {
-      continue;
-    } else {
-      newUrl.push(oldUrl[i]);
-    }
-  }
-
-  return newUrl.join("/");
-}
-
-/**
- * Construct an url from the arguments given.
- * Basically:
- *   - The last arguments that contains a scheme (e.g. "http://") is the base
- *     of the url.
- *   - every subsequent string arguments are concatened to it.
- * @param {...string|undefined} args
- * @returns {string}
- */
-export default function resolveURL(...args: Array<string | undefined>): string {
-  const len = args.length;
-  if (len === 0) {
-    return "";
-  }
-
-  let base = "";
-  for (let i = 0; i < len; i++) {
-    let part = args[i];
-    if (typeof part !== "string" || part === "") {
-      continue;
-    }
-    if (schemeRe.test(part)) {
-      base = part;
-    } else {
-      // trim if begins with "/"
-      if (part[0] === "/") {
-        part = part.substring(1);
-      }
-
-      // trim if ends with "/"
-      if (base[base.length - 1] === "/") {
-        base = base.substring(0, base.length - 1);
-      }
-
-      base = base + "/" + part;
-    }
-  }
-
-  return _normalizeUrl(base);
-}
+/** 
+ * Match the different components of an URL.
+ * 
+ *     foo://example.com:8042/over/there?name=ferret#nose
+       \_/   \______________/\_________/ \_________/ \__/
+        |           |            |            |        |
+      scheme     authority       path        query   fragment
+ * 1st match is the scheme: (e.g. "foo://")
+ * 2nd match is the authority (e.g "example.com:8042")
+ * 3rd match is the path (e.g "/over/there")
+ * 4th match is the query params (e.g "name=ferret")
+ * 5th match is the fragment (e.g "nose")
+ * */
+const urlComponentRegex =
+  /^(?:([^:\/?#]+):)?(?:\/\/([^\/?#]*))?([^?#]*)(?:\?([^#]*))?(?:#(.*))?$/;
```

Comment:
Damn, Where'd You Find This?

Reply:
This is a version of this regex https://github.com/google/closure-library/blob/master/closure/goog/uri/utils.js#L189
without the userInfo & port, it thought it was unnecessary. 
Maybe having splitting it to explain every part can be good for readability purpose

Reply:
OK nice, I thought that this was a pretty common regexp so I asked.

>  without the userInfo & port, it thought it was unnecessary.
What if one of those is present (especially the port)? It should still work, just not exploited, no?

Reply:
If there is a port or user credentials, it's captured within the "authority" part of the regex.
`http://username:password@example.com:8090/foo/`
===>
scheme: `http`
authority: `username:password@example.com:8090`
path: `/foo`

The port and user credentials are not parsed into there own component but here we don't need to extract them individually.

I am adding some unit tests to ensure that this behavior is well supported.